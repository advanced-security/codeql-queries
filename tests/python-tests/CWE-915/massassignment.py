from flask import Flask, request, abort

app = Flask(__name__)


class Users(object):
    uid: int
    username: str
    password: str


@app.route("/test1")
def test1():
    # Test 1: Simple test case
    i = request.args.get("i")
    user = Users()
    
    # codeql: py/mass-assignment
    setattr(user, request.args.get("k"), request.args.get("v"))
    
    # codeql: py/mass-assignment
    user.__setattr__(request.args.get("k"), request.args.get("v"))

    # false-positive: variable isn't user controlled
    setattr(user, "uid", request.args.get("uid"))
    # same as: user.uid = i
    
    # false-positive: variable isn't user controlled
    user.__setattr__("uid", request.args.get("uid"))
    # same as: user.uid = i
    
    return f"<h1>Test 3</h1>"


@app.route("/test3")
def test3():
    user = Users()
    # Test 2:
    # remote-flow: requrst.args
    for param_name, param_vaule in request.args.items():
        # codeql: py/mass-assignment
        setattr(user, param_name, param_vaule)
    
    return f"<h1>Test 3</h1>"


@app.route("/test4")
def test4():
    # Test 3:
    if request.method == "PUT":
        if not request.json:
                abort(400)

        user_data = request.json
        user = Users()

        # remote-flow: requrst.args
        for item in user_data:
            # codeql: py/mass-assignment
            setattr(user, item, request.json[item])

    return f"<h1>Test 4</h1>"
