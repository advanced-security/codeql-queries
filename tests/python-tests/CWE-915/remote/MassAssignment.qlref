CWE-915/MassAssignment.ql