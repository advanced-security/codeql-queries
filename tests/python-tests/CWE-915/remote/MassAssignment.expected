edges
| massassignment.py:11:19:11:21 | ControlFlowNode for key | massassignment.py:13:26:13:28 | ControlFlowNode for key |
| massassignment.py:19:9:19:15 | ControlFlowNode for request | massassignment.py:23:19:23:30 | ControlFlowNode for Attribute |
| massassignment.py:19:9:19:15 | ControlFlowNode for request | massassignment.py:26:22:26:33 | ControlFlowNode for Attribute |
| massassignment.py:19:9:19:15 | ControlFlowNode for request | massassignment.py:29:14:29:25 | ControlFlowNode for Attribute |
| massassignment.py:23:19:23:25 | ControlFlowNode for request | massassignment.py:23:19:23:30 | ControlFlowNode for Attribute |
| massassignment.py:23:19:23:25 | ControlFlowNode for request | massassignment.py:26:22:26:33 | ControlFlowNode for Attribute |
| massassignment.py:23:19:23:25 | ControlFlowNode for request | massassignment.py:29:14:29:25 | ControlFlowNode for Attribute |
| massassignment.py:23:19:23:30 | ControlFlowNode for Attribute | massassignment.py:23:19:23:39 | ControlFlowNode for Attribute() |
| massassignment.py:23:42:23:48 | ControlFlowNode for request | massassignment.py:26:22:26:33 | ControlFlowNode for Attribute |
| massassignment.py:23:42:23:48 | ControlFlowNode for request | massassignment.py:29:14:29:25 | ControlFlowNode for Attribute |
| massassignment.py:26:22:26:28 | ControlFlowNode for request | massassignment.py:26:22:26:33 | ControlFlowNode for Attribute |
| massassignment.py:26:22:26:28 | ControlFlowNode for request | massassignment.py:29:14:29:25 | ControlFlowNode for Attribute |
| massassignment.py:26:22:26:33 | ControlFlowNode for Attribute | massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() |
| massassignment.py:26:45:26:51 | ControlFlowNode for request | massassignment.py:29:14:29:25 | ControlFlowNode for Attribute |
| massassignment.py:29:14:29:20 | ControlFlowNode for request | massassignment.py:29:14:29:25 | ControlFlowNode for Attribute |
| massassignment.py:29:14:29:25 | ControlFlowNode for Attribute | massassignment.py:29:14:29:34 | ControlFlowNode for Attribute() |
| massassignment.py:29:14:29:34 | ControlFlowNode for Attribute() | massassignment.py:11:19:11:21 | ControlFlowNode for key |
| massassignment.py:47:9:47:18 | SSA variable param_name | massassignment.py:49:23:49:32 | ControlFlowNode for param_name |
| massassignment.py:47:9:47:31 | ControlFlowNode for Tuple | massassignment.py:47:9:47:18 | SSA variable param_name |
| massassignment.py:47:9:47:31 | ControlFlowNode for Tuple [Tuple element at index 0] | massassignment.py:47:9:47:18 | SSA variable param_name |
| massassignment.py:47:9:47:31 | IterableElement | massassignment.py:47:9:47:31 | ControlFlowNode for Tuple |
| massassignment.py:47:9:47:31 | IterableElement | massassignment.py:47:9:47:31 | ControlFlowNode for Tuple [Tuple element at index 0] |
| massassignment.py:47:9:47:31 | IterableSequence | massassignment.py:47:9:47:31 | ControlFlowNode for Tuple |
| massassignment.py:47:9:47:31 | IterableSequence | massassignment.py:47:9:47:31 | IterableElement |
| massassignment.py:47:36:47:42 | ControlFlowNode for request | massassignment.py:47:36:47:47 | ControlFlowNode for Attribute |
| massassignment.py:47:36:47:47 | ControlFlowNode for Attribute | massassignment.py:47:9:47:31 | IterableSequence |
| massassignment.py:57:8:57:14 | ControlFlowNode for request | massassignment.py:61:21:61:32 | ControlFlowNode for Attribute |
| massassignment.py:58:16:58:22 | ControlFlowNode for request | massassignment.py:61:21:61:32 | ControlFlowNode for Attribute |
| massassignment.py:61:21:61:27 | ControlFlowNode for request | massassignment.py:61:21:61:32 | ControlFlowNode for Attribute |
| massassignment.py:61:21:61:32 | ControlFlowNode for Attribute | massassignment.py:65:13:65:16 | SSA variable item |
| massassignment.py:65:13:65:16 | SSA variable item | massassignment.py:67:27:67:30 | ControlFlowNode for item |
nodes
| massassignment.py:11:19:11:21 | ControlFlowNode for key | semmle.label | ControlFlowNode for key |
| massassignment.py:13:26:13:28 | ControlFlowNode for key | semmle.label | ControlFlowNode for key |
| massassignment.py:19:9:19:15 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:23:19:23:25 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:23:19:23:30 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:23:19:23:39 | ControlFlowNode for Attribute() | semmle.label | ControlFlowNode for Attribute() |
| massassignment.py:23:42:23:48 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:26:22:26:28 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:26:22:26:33 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | semmle.label | ControlFlowNode for Attribute() |
| massassignment.py:26:45:26:51 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:29:14:29:20 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:29:14:29:25 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:29:14:29:34 | ControlFlowNode for Attribute() | semmle.label | ControlFlowNode for Attribute() |
| massassignment.py:47:9:47:18 | SSA variable param_name | semmle.label | SSA variable param_name |
| massassignment.py:47:9:47:31 | ControlFlowNode for Tuple | semmle.label | ControlFlowNode for Tuple |
| massassignment.py:47:9:47:31 | ControlFlowNode for Tuple [Tuple element at index 0] | semmle.label | ControlFlowNode for Tuple [Tuple element at index 0] |
| massassignment.py:47:9:47:31 | IterableElement | semmle.label | IterableElement |
| massassignment.py:47:9:47:31 | IterableSequence | semmle.label | IterableSequence |
| massassignment.py:47:36:47:42 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:47:36:47:47 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:49:23:49:32 | ControlFlowNode for param_name | semmle.label | ControlFlowNode for param_name |
| massassignment.py:57:8:57:14 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:58:16:58:22 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:61:21:61:27 | ControlFlowNode for request | semmle.label | ControlFlowNode for request |
| massassignment.py:61:21:61:32 | ControlFlowNode for Attribute | semmle.label | ControlFlowNode for Attribute |
| massassignment.py:65:13:65:16 | SSA variable item | semmle.label | SSA variable item |
| massassignment.py:67:27:67:30 | ControlFlowNode for item | semmle.label | ControlFlowNode for item |
subpaths
#select
| massassignment.py:13:26:13:28 | ControlFlowNode for key | massassignment.py:19:9:19:15 | ControlFlowNode for request | massassignment.py:13:26:13:28 | ControlFlowNode for key | Use of $@. | massassignment.py:19:9:19:15 | ControlFlowNode for request | mass assignment |
| massassignment.py:13:26:13:28 | ControlFlowNode for key | massassignment.py:23:19:23:25 | ControlFlowNode for request | massassignment.py:13:26:13:28 | ControlFlowNode for key | Use of $@. | massassignment.py:23:19:23:25 | ControlFlowNode for request | mass assignment |
| massassignment.py:13:26:13:28 | ControlFlowNode for key | massassignment.py:23:42:23:48 | ControlFlowNode for request | massassignment.py:13:26:13:28 | ControlFlowNode for key | Use of $@. | massassignment.py:23:42:23:48 | ControlFlowNode for request | mass assignment |
| massassignment.py:13:26:13:28 | ControlFlowNode for key | massassignment.py:26:22:26:28 | ControlFlowNode for request | massassignment.py:13:26:13:28 | ControlFlowNode for key | Use of $@. | massassignment.py:26:22:26:28 | ControlFlowNode for request | mass assignment |
| massassignment.py:13:26:13:28 | ControlFlowNode for key | massassignment.py:26:45:26:51 | ControlFlowNode for request | massassignment.py:13:26:13:28 | ControlFlowNode for key | Use of $@. | massassignment.py:26:45:26:51 | ControlFlowNode for request | mass assignment |
| massassignment.py:13:26:13:28 | ControlFlowNode for key | massassignment.py:29:14:29:20 | ControlFlowNode for request | massassignment.py:13:26:13:28 | ControlFlowNode for key | Use of $@. | massassignment.py:29:14:29:20 | ControlFlowNode for request | mass assignment |
| massassignment.py:23:19:23:39 | ControlFlowNode for Attribute() | massassignment.py:19:9:19:15 | ControlFlowNode for request | massassignment.py:23:19:23:39 | ControlFlowNode for Attribute() | Use of $@. | massassignment.py:19:9:19:15 | ControlFlowNode for request | mass assignment |
| massassignment.py:23:19:23:39 | ControlFlowNode for Attribute() | massassignment.py:23:19:23:25 | ControlFlowNode for request | massassignment.py:23:19:23:39 | ControlFlowNode for Attribute() | Use of $@. | massassignment.py:23:19:23:25 | ControlFlowNode for request | mass assignment |
| massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | massassignment.py:19:9:19:15 | ControlFlowNode for request | massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | Use of $@. | massassignment.py:19:9:19:15 | ControlFlowNode for request | mass assignment |
| massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | massassignment.py:23:19:23:25 | ControlFlowNode for request | massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | Use of $@. | massassignment.py:23:19:23:25 | ControlFlowNode for request | mass assignment |
| massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | massassignment.py:23:42:23:48 | ControlFlowNode for request | massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | Use of $@. | massassignment.py:23:42:23:48 | ControlFlowNode for request | mass assignment |
| massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | massassignment.py:26:22:26:28 | ControlFlowNode for request | massassignment.py:26:22:26:42 | ControlFlowNode for Attribute() | Use of $@. | massassignment.py:26:22:26:28 | ControlFlowNode for request | mass assignment |
| massassignment.py:49:23:49:32 | ControlFlowNode for param_name | massassignment.py:47:36:47:42 | ControlFlowNode for request | massassignment.py:49:23:49:32 | ControlFlowNode for param_name | Use of $@. | massassignment.py:47:36:47:42 | ControlFlowNode for request | mass assignment |
| massassignment.py:67:27:67:30 | ControlFlowNode for item | massassignment.py:57:8:57:14 | ControlFlowNode for request | massassignment.py:67:27:67:30 | ControlFlowNode for item | Use of $@. | massassignment.py:57:8:57:14 | ControlFlowNode for request | mass assignment |
| massassignment.py:67:27:67:30 | ControlFlowNode for item | massassignment.py:58:16:58:22 | ControlFlowNode for request | massassignment.py:67:27:67:30 | ControlFlowNode for item | Use of $@. | massassignment.py:58:16:58:22 | ControlFlowNode for request | mass assignment |
| massassignment.py:67:27:67:30 | ControlFlowNode for item | massassignment.py:61:21:61:27 | ControlFlowNode for request | massassignment.py:67:27:67:30 | ControlFlowNode for item | Use of $@. | massassignment.py:61:21:61:27 | ControlFlowNode for request | mass assignment |
